### ISSUE TITLE
Rewrite the regularize step

### ISSUE THREAD

**Issue by kordc** [SebChw/Actually-Robust-Training#177]

**Comment by kordc** [comment]
Regularizations from Karpathy:

4. Regularize
Ideally, we are now at a place where we have a large model that is fitting at least the training set. Now it is time to regularize it and gain some validation accuracy by giving up some of the training accuracy. Some tips & tricks:

get more data. First, the by far best and preferred way to regularize a model in any practical setting is to add more real training data. It is a very common mistake to spend a lot engineering cycles trying to squeeze juice out of a small dataset when you could instead be collecting more data. As far as I’m aware adding more data is pretty much the only guaranteed way to monotonically improve the performance of a well-configured neural network almost indefinitely. The other would be ensembles (if you can afford them), but that tops out after ~5 models.
data augment. The next best thing to real data is half-fake data - try out more aggressive data augmentation.
creative augmentation. If half-fake data doesn’t do it, fake data may also do something. People are finding creative ways of expanding datasets; For example, [domain randomization](https://openai.com/blog/learning-dexterity/), use of [simulation](http://vladlen.info/publications/playing-data-ground-truth-computer-games/), clever [hybrids](https://arxiv.org/abs/1708.01642) such as inserting (potentially simulated) data into scenes, or even GANs.
pretrain. It rarely ever hurts to use a pretrained network if you can, even if you have enough data.
stick with supervised learning. Do not get over-excited about unsupervised pretraining. Unlike what that blog post from 2008 tells you, as far as I know, no version of it has reported strong results in modern computer vision (though NLP seems to be doing pretty well with BERT and friends these days, quite likely owing to the more deliberate nature of text, and a higher signal to noise ratio).
smaller input dimensionality. Remove features that may contain spurious signal. Any added spurious input is just another opportunity to overfit if your dataset is small. Similarly, if low-level details don’t matter much try to input a smaller image.
smaller model size. In many cases you can use domain knowledge constraints on the network to decrease its size. As an example, it used to be trendy to use Fully Connected layers at the top of backbones for ImageNet but these have since been replaced with simple average pooling, eliminating a ton of parameters in the process.
decrease the batch size. Due to the normalization inside batch norm smaller batch sizes somewhat correspond to stronger regularization. This is because the batch empirical mean/std are more approximate versions of the full mean/std so the scale & offset “wiggles” your batch around more.
drop. Add dropout. Use dropout2d (spatial dropout) for ConvNets. Use this sparingly/carefully because dropout [does not seem to play nice](https://arxiv.org/abs/1801.05134) with batch normalization.
weight decay. Increase the weight decay penalty.
early stopping. Stop training based on your measured validation loss to catch your model just as it’s about to overfit.
try a larger model. I mention this last and only after early stopping but I’ve found a few times in the past that larger models will of course overfit much more eventually, but their “early stopped” performance can often be much better than that of smaller models.
Finally, to gain additional confidence that your network is a reasonable classifier, I like to visualize the network’s first-layer weights and ensure you get nice edges that make sense. If your first layer filters look like noise then something could be off. Similarly, activations inside the net can sometimes display odd artifacts and hint at problems.

**Comment by SebChw** [comment]
Also my thoughts on this

Regularization step - The goal of this step is to help the user track/use techniques used for regularizations and their result

What currently `Regularize` step gives to the user:
1. Information that this run of training was all about Regularization

Expected Pros of using `Regularize` step:
1. Information on what techniques was already used and what was the results
2. Some helper functions for regularization
3. Suggestions on what more could be used.
4. Easy reproduction for different settings
5. Easy comparison and progress tracking
6. no messy configs or constant need for code changes.

Current Implementation:
1. Every model implements turn_on_regularization function, every dataset respectively implements the same function:
   1. Pros
      1. Simple
      2. Everything on the user
   2. Cons
      1. Cumbersome for user to implement as it doesn't have access for model/dataset initialization where many regularization things are set up
      2. Validates Open/Closed principle
      3. Hard to log information from it. We would need another log_regularization etc.

Ideas:
1. User defines default_model_setting() for every created model
   1. Pros
      1. we know what was changed and we can create somple plots as stuff changes (learning curves)
   2. Cons
      1. User may forget to update it.
      2. Magic values won't be tracked.
2. We force user to check whether defined parameter changes model regularization or not.
   1. Pros
      1. Simple
   2. Cons
      1. Everthing is put on the user, error prone as user can forget to update it.
3. Implement steps/functions that apply different regularization techniques and user just uses them.
   1. Pros
      1. Not user dependent, very simple to check if we use it.
      2. Information about which techniques was already used can be easily displayed
   2. cons
      1. Not all are possible to implement
      2. Even if possible to implement probably would be very hacky.
4. Add some functions to art_module, art_dataset that user must implement and return False/logged value to indicate if they use certain regularization.
   1. Cons
      1. Even if someone uses just 1 regularization they must implement this function for every technique.
5. By default create other hidden git repository that tracks every run experiment and connects a comit with it.
   1. Pros
      1. Almost always allows for reproduction
   2. Cons
      1. Only local

Possible regularization techniques and whether they are trackable:
1. Getting more data - This can be easily detected by inspecting length of train dataset inside train dataloader.
2. Data augment - Done in dataset most often by setting some `transform` function. But user may call it whatever they want.
3. Pretrain - basically no way to detect this automatically. Beside user passing ckpt path in trainer or
4. Smaller input dimensionality/resolution - detectable but cumbersome, many assumptions must be done
5. dropout - again detectable but how to know if user changed this?
6. weight decay - -||-
7. Early stopping - same story
8. many more...

**Comment by SebChw** [comment]
How I see it now - it should just expect list of configurations + list of lists of model modifiers and run many trainings.

Moreover I'd consider adding something like datamodule modifier?

**Comment by kordc** [comment]
I like this idea. The datamodule modifier could be useful as we define this step as a sequence of trainings.
